test/test_upload.py

```diff
@@ -195,3 +198,143 @@ def test_compile_and_upload_combo_with_custom_build_path(run_command, data_dir,
+
+        # Compile sketch first
+        assert run_command(f"compile --clean -b {board.fqbn} {sketch_path}")
+
+        # Upload from sketch folder
+        wait_for_board()
+        assert run_command(f"upload -b {board.fqbn} -p {board.address} {sketch_path}")
+
+        # Upload from sketch file
+        wait_for_board()
+        assert run_command(f"upload -b {board.fqbn} -p {board.address} {sketch_file}")
+
+        # Upload from build folder
+        sketch_path_md5 = hashlib.md5(bytes(sketch_path)).hexdigest().upper()
+        build_dir = Path(tempfile.gettempdir(), f"arduino-sketch-{sketch_path_md5}")
+        wait_for_board()
+        res = run_command(f"upload -b {board.fqbn} -p {board.address} --input-dir {build_dir}")
+        assert (
+            "Sketches with .pde extension are deprecated, please rename the following files to .ino:" not in res.stderr
+        )
+
+        # Upload from binary file
+        wait_for_board()
+        # We don't need a specific file when using the --input-file flag to upload since
+        # it's just used to calculate the directory, so it's enough to get a random file
+        # that's inside that directory
+        binary_file = next(build_dir.glob(f"{sketch_name}.pde.*"))
+        res = run_command(f"upload -b {board.fqbn} -p {board.address} --input-file {binary_file}")
+        assert (
+            "Sketches with .pde extension are deprecated, please rename the following files to .ino:" not in res.stderr
+        )
+
+
+def test_upload_with_input_dir_containing_multiple_binaries(run_command, data_dir, detected_boards, wait_for_board):
+    # This tests verifies the behaviour outlined in this issue:
+    # https://github.com/arduino/arduino-cli/issues/765#issuecomment-699678646
+    assert run_command("update")
+
+    # Create a two different sketches
+    sketch_one_name = "UploadMultipleBinariesSketchOne"
+    sketch_one_path = Path(data_dir, sketch_one_name)
+    assert run_command(f"sketch new {sketch_one_path}")
+
+    sketch_two_name = "UploadMultipleBinariesSketchTwo"
+    sketch_two_path = Path(data_dir, sketch_two_name)
+    assert run_command(f"sketch new {sketch_two_path}")
+
+    for board in detected_boards:
+        # Install core
+        core = ":".join(board.fqbn.split(":")[:2])
+        assert run_command(f"core install {core}")
+
+        # Compile both sketches and copy binaries in the same directory same build directory
+        assert run_command(f"compile --clean -b {board.fqbn} {sketch_one_path}")
+        assert run_command(f"compile --clean -b {board.fqbn} {sketch_two_path}")
+
+        sketch_path_md5 = hashlib.md5(bytes(sketch_one_path)).hexdigest().upper()
+        build_dir_one = Path(tempfile.gettempdir(), f"arduino-sketch-{sketch_path_md5}")
+
+        sketch_path_md5 = hashlib.md5(bytes(sketch_two_path)).hexdigest().upper()
+        build_dir_two = Path(tempfile.gettempdir(), f"arduino-sketch-{sketch_path_md5}")
```

Comment:
again, build path should be get from `compile --format json`, there may be other places to change